src/mrpro/operators/_CartesianSamplingOp.py

```diff
@@ -0,0 +1,172 @@
+                kidx.to(dtype=torch.int64, device=traj.kx.device), 'other k2 k1 k0->other coil k2 k1 k0', coil=1
+            )
+
+            self._fft_idx = kidx
+            self._fft_idx_full = torch.zeros(0)  # not None to satisfy mypy
+
+        # Make sure sorting index is actually needed
+        if torch.all(torch.diff(self._fft_idx.flatten()) == 1):
+            self._fft_idx = torch.zeros(0)  # not None to satisfy mypy
+
+        self._fft_dims = tuple(fft_dims)
+        self._kshape = traj.broadcasted_shape
+        self._encoding_shape = encoding_shape
+
+    def forward(self, y: torch.Tensor) -> torch.Tensor:
+        """Forward operator which selects acquired k-space data from k-space.
+
+        Parameters
+        ----------
+        y
+            k-space with dimensions given by encoding_shape
+
+        Returns
+        -------
+            k-space data in original (i.e. acquired) dimensions
+        """
+        if self._encoding_shape != SpatialDimension(*y.shape[-3:]):
+            raise ValueError('k-space data shape missmatch')
+
+        if len(self._fft_dims) > 0 and len(self._fft_idx) > 0:
+            if len(self._fft_idx_full) == 0 or self._fft_idx_full.shape[1] != y.shape[1]:
+                coil_idx = torch.ones(
+                    *(y.shape[:2] + self._fft_idx.shape[2:]), dtype=torch.int64, device=y.device
+                ) * rearrange(
+                    torch.linspace(0, y.shape[1] - 1, y.shape[1], dtype=torch.int64, device=y.device),
+                    'coils->1 coils 1 1 1',
+                )
+                other_idx = torch.ones_like(self._fft_idx, dtype=torch.int64, device=y.device) * rearrange(
+                    torch.linspace(0, y.shape[0] - 1, y.shape[0], dtype=torch.int64, device=y.device),
+                    'other->other 1 1 1 1',
+                )
+                self._fft_idx_full = (
+                    self._fft_idx
+                    + other_idx * y.shape[1]
+                    + coil_idx * self._encoding_shape.z * self._encoding_shape.y * self._encoding_shape.x
+                )
+
+            return torch.take(y, self._fft_idx_full)
+        else:
+            return y
+
+    def adjoint(self, y: torch.Tensor) -> torch.Tensor:
+        """Adjoint operator sorting data into the encoding_space matrix.
+
+        Parameters
+        ----------
+        y
+            k-space data in original (i.e. acquired shape)
+
+        Returns
+        -------
+            k-space data sorted into encoding_space matrix
+        """
+
+        if self._kshape[1:] != y.shape[-3:]:
+            raise ValueError('k-space data shape missmatch')
+
+        if len(self._fft_dims) > 0 and len(self._fft_idx) > 0:
+            if len(self._fft_idx_full) == 0 or self._fft_idx_full.shape[1] != y.shape[1]:
+                coil_idx = torch.ones(
+                    *(y.shape[:2] + self._fft_idx.shape[2:]), dtype=torch.int64, device=y.device
+                ) * rearrange(
```

Comment:
is this the same as 
`einops.repeat(torch.arange(y.shape[-4], device=y.device), " coils-> other coils k2 k1 k0", k0=y.shape[-1], ...  `
i.e. 
arrange instead of linspace, 
using einops repeat instead of reanage and multiplying by ones?

Reply:
nice